src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs

```diff
@@ -141,50 +147,46 @@ protected virtual void ReadKeyDescriptorForSigning(WsFederationConfiguration con
+
+            KeyInfo keyInfo = null;
 
             if (reader.IsStartElement(XmlSignatureConstants.Elements.KeyInfo, XmlSignatureConstants.Namespace))
-            {
-                var keyInfo = _dsigSerializer.ReadKeyInfo(reader);
-                configuration.KeyInfos.Add(keyInfo);
-                if (!string.IsNullOrEmpty(keyInfo.CertificateData))
-                {
-                    var cert = new X509Certificate2(Convert.FromBase64String(keyInfo.CertificateData));
-                    configuration.SigningKeys.Add(new X509SecurityKey(cert));
-                }
-            }
+                keyInfo = _dsigSerializer.ReadKeyInfo(reader);
             else
-            {
                 throw XmlUtil.LogReadException(LogMessages.IDX13002, reader.LocalName, reader.NamespaceURI, XmlSignatureConstants.Elements.KeyInfo, XmlSignatureConstants.Namespace);
-            }
 
             // </KeyDescriptor>
             reader.ReadEndElement();
+
+            if (keyInfo == null)
+                throw XmlUtil.LogReadException(LogMessages.IDX13010);
+
+            return keyInfo;
         }
 
         /// <summary>
         /// Read RoleDescriptor element in xml.
         /// </summary>
-        /// <param name="configuration">WsFed configuration</param>
         /// <param name="reader">xmlreader</param>
-        protected virtual void ReadSecurityTokenServiceTypeRoleDescriptor(WsFederationConfiguration configuration, XmlReader reader)
+        /// <returns>Security token service type role descriptor</returns>
+        protected virtual SecurityTokenServiceTypeRoleDescriptor ReadSecurityTokenServiceTypeRoleDescriptor(XmlReader reader)
         {
-            if (configuration == null)
-                throw LogArgumentNullException(nameof(configuration));
-
             XmlUtil.CheckReaderOnEntry(reader, Elements.RoleDescriptor, Namespaces.MetadataNamespace);
 
             if (!IsSecurityTokenServiceTypeRoleDescriptor(reader))
                 throw XmlUtil.LogReadException(LogMessages.IDX13004);
 
+            var roleDescriptor = new SecurityTokenServiceTypeRoleDescriptor();
+
             // <RoleDescriptorr>
             reader.ReadStartElement();
             while (reader.IsStartElement())
             {
                 if (reader.IsStartElement(Elements.KeyDescriptor, Namespaces.MetadataNamespace) && reader.GetAttribute(Attributes.Use).Equals(keyUse.Signing))
-                    ReadKeyDescriptorForSigning(configuration, reader);
+                    roleDescriptor.KeyInfos.Add(ReadKeyDescriptorForSigning(reader));
                 else if (reader.IsStartElement(Elements.SecurityTokenEndpoint, Namespaces.FederationNamespace))
-                    ReadSecurityTokenEndpoint(configuration, reader);
+                    roleDescriptor.TokenEndpoint = ReadSecurityTokenEndpoint(reader);
                 else if (reader.IsStartElement())
                     reader.ReadOuterXml();
                 else
```

Comment:
LogMessages.IDX13003 doesn't describe this exception completely. 
It should more descriptive.